### PR TITLE
timer: add support for sigev_thread flag in zephyr timer posix api

### PR DESCRIPTION
**Pull request by arun-mani-tech**
add support for the sigev_thread flag in zephyr timer posix api, following the behavior described in the linux man page. with this enhancement, a single thread is created to receive all notifications from timers

### CODE REVIEW

**Comment by cfriedt** [comment]
This is probably the specification we should refer to in this pr

https://pubs.opengroup.org/onlinepubs/9699919799/functions/timer_create.html

File: lib/posix/timer.c
Comment (cfriedt) [inline lib/posix/timer.c:284]: The spec explicitly says that the threads should be created detached. Therefore we should not call join.

File: lib/posix/timer.c
Comment (cfriedt) [inline lib/posix/timer.c:111]: There is no need to manually allocate a thread stack in this function.

Thread stack allocation is automatic with `pthread_create(&tid, NULL, func, arg)`. I.e. with a `NULL` `pthread_attr_t`.

Comment (cfriedt) [reply lib/posix/timer.c:111]: The spec says that users may pass in a `pthread_attr_t *` may be passed in via `sigev_notify_attributes` and if it is non-`NULL` then we should use the stack provided from those attributes.

File: lib/posix/timer.c
Comment (cfriedt) [inline lib/posix/timer.c:25]: This is unnecessary, since it is already contained within the `pthread_attr_t`

File: lib/posix/timer.c
Comment (cfriedt) [inline lib/posix/timer.c:24]: This is unecessary to keep, and can be discarded after `pthread_create()` is called

**Review by cfriedt — CHANGES_REQUESTED** [review]
There seems to be some misaligned from the spec. That needs to be addressed before it is considered.

Some other issues that seem unnecessary (unnecessary storage of things in a structure, unnecessary allocation, etc).

Please also create additional tests for the testsuite as a separate second commit to demonstrate and verify functionality.

**Comment by cfriedt** [comment]
@arun-mani-tech - can you please rebase on latest `main` and resolve conflicts?

The tests that are failing are unrelated and passing on `main`

**Comment by cfriedt** [comment]
@arun-mani-tech - you may have better luck using pool allocated thread stacks rather than heap allocated thread stacks. Heap allocation can require more memory resources than what is available on certain platforms to satisfy alignments, MPU, etc.

See e.g. #60759 

```
CONFIG_POSIX_API=y
CONFIG_THREAD_STACK_INFO=y
CONFIG_DYNAMIC_THREAD=y
CONFIG_DYNAMIC_THREAD_POOL_SIZE=2
```

File: lib/posix/timer.c
Comment (cfriedt) [inline lib/posix/timer.c:114]:
```suggestion
		ret = pthread_create(&timer->thread, evp->sigev_notify_attributes, zephyr_thread_wrapper, timer);
```

**Comment by cfriedt** [comment]
@arun-mani-tech - you may have better luck using pool allocated thread stacks rather than heap allocated thread stacks. Heap allocation can require more memory resources than what is available on certain platforms to satisfy alignments, MPU, etc.

See e.g. #60759 

```
CONFIG_POSIX_API=y
CONFIG_THREAD_STACK_INFO=y
CONFIG_DYNAMIC_THREAD=y
CONFIG_DYNAMIC_THREAD_POOL_SIZE=2
```

**Review by cfriedt — CHANGES_REQUESTED** [review]
Minor change still needed for the call to `pthread_create()` is to use the caller provided `sigev_notify_attributes`. If it is `NULL`, then a thread stack is automatically allocated. 

Otherwise, looks good. I would maybe use pooled / dynamic thread stacks since they are a bit easier on arches with complicated MPU alignments.

**Comment by cfriedt** [comment]
One of the test failures leads me to believe that we should be filtering out `arch_posix` in a `testcase.yaml` file.

```
2023-09-08T14:56:32.5934159Z INFO    - -+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
2023-09-08T14:56:32.5934859Z INFO    - The following issues were found (showing the top 10 items):
2023-09-08T14:56:32.5935559Z INFO    - 1) tests/posix/common/portability.posix.common.picolibc on nrf52_bsim failed (Failed)
2023-09-08T14:56:32.5935994Z INFO    - 
2023-09-08T14:56:32.5936484Z INFO    - To rerun the tests, call twister using the following commandline:
2023-09-08T14:56:32.5937047Z INFO    - west twister -p <PLATFORM> -s <TEST ID>, for example:
2023-09-08T14:56:32.5937443Z INFO    - 
2023-09-08T14:56:32.5937929Z INFO    - west twister -p nrf52_bsim -s tests/posix/common/portability.posix.common.picolibc
2023-09-08T14:56:32.5938363Z INFO    - or with west:

```

**Comment by cfriedt** [comment]
it seems to fail on `mps2_an385`
```
In file included from /__w/zephyr/zephyr/lib/posix/timer.c:9:
2023-09-08T14:33:56.1660296Z /__w/zephyr/zephyr/include/zephyr/posix/pthread.h:445:18: error: unknown type name 'pthread_once_t'; did you mean 'pthread_cond_t'?
2023-09-08T14:33:56.1660840Z   445 | int pthread_once(pthread_once_t *once, void (*initFunc)(void));
2023-09-08T14:33:56.1661195Z       |                  ^~~~~~~~~~~~~~
```

**Comment by cfriedt** [comment]
it seems to fail on `mps2_an385`
```
In file included from /__w/zephyr/zephyr/lib/posix/timer.c:9:
2023-09-08T14:33:56.1660296Z /__w/zephyr/zephyr/include/zephyr/posix/pthread.h:445:18: error: unknown type name 'pthread_once_t'; did you mean 'pthread_cond_t'?
2023-09-08T14:33:56.1660840Z   445 | int pthread_once(pthread_once_t *once, void (*initFunc)(void));
2023-09-08T14:33:56.1661195Z       |                  ^~~~~~~~~~~~~~
```

**Comment by arun-mani-tech** [comment]
@cfriedt Changes have been made please revisit

**Review by cfriedt — APPROVED** [review]

**Comment by arun-mani-tech** [comment]
@ycsin please review

**Comment by arun-mani-tech** [comment]
@povergoing @jgl-meta Please review this

**Review by ycsin — APPROVED** [review]

**Comment by ycsin** [comment]
Not a fan of single line if-conditionals without curly braces, otherwise LGTM

**Review by povergoing — APPROVED** [review]
Overall LGTM, also not a fan of if-conditionals without curly braces

**Pull request merged**
